File: src/main/java/com/frank/entity/courier/strategies/impl/NearestDeliveryStrategy.java
package com.frank.entity.courier.strategies.impl;

import com.frank.entity.order.Order;
import com.frank.entity.courier.Courier;
import com.frank.entity.courier.strategies.BestDeliveryStrategy;
import com.frank.entity.courier.dto.ShipmentDetails;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class NearestDeliveryStrategy implements BestDeliveryStrategy {

    final private List<Courier> couriers = new ArrayList<>();

    final private Order order;

    public NearestDeliveryStrategy(Order order, Courier... couriers) {
        this.order = order;
        this.couriers.addAll(Arrays.asList(couriers));
    }

    @Override
    public ShipmentDetails calculate() {
        ShipmentDetails shipmentDetails = new ShipmentDetails();
        if (this.couriers.size() > 0) {
            Courier best = this.couriers.get(0);
            for (Courier courier : this.couriers) {
                if (getMinDateAmongList(courier.getAvailability()).isBefore(getMinDateAmongList(best.getAvailability()))) {
                    best = courier;
                }
            }
            shipmentDetails.setCourier(best);
        }
        return shipmentDetails;
    }

    private LocalDate getMinDateAmongList(List<LocalDate> list) {
        return list.stream().min(LocalDate::compareTo).get();
    }
}

File: src/main/java/com/frank/entity/courier/strategies/impl/NearestDeliveryStrategy.kt
package com.frank.entity.courier.strategies.impl

import com.frank.entity.order.Order
import com.frank.entity.courier.Courier
import com.frank.entity.courier.strategies.BestDeliveryStrategy
import com.frank.entity.courier.dto.ShipmentDetails

import java.time.LocalDate
import java.util.ArrayList
import java.util.Arrays

class NearestDeliveryStrategy(private val order: Order, vararg couriers: Courier) : BestDeliveryStrategy {

    private val couriers = ArrayList<Courier>()

    init {
        this.couriers.addAll(Arrays.asList(*couriers))
    }

    override fun calculate(): ShipmentDetails {
        val shipmentDetails = ShipmentDetails()
        if (this.couriers.size > 0) {
            var best = this.couriers[0]
            for (courier in this.couriers) {
                if (getMinDateAmongList(courier.availability!!).isBefore(getMinDateAmongList(best.availability!!))) {
                    best = courier
                }
            }
            shipmentDetails.courier = best
        }
        return shipmentDetails
    }

    private fun getMinDateAmongList(list: List<LocalDate>): LocalDate {
        return list.stream().min { obj, other -> obj.compareTo(other) }.get()
    }
}
